src/ui/utils/sources.ts

```diff
@@ -86,36 +100,40 @@ export const newSourcesToCompleteSourceDetails = (
     return `${source.kind}:${source.url!}:${contentHashForSource(source)}`;
   };
 
-  newSources.forEach(source => {
+  for (let source of newSources) {
     const key = keyForSource(source);
     if (corresponding[key] === undefined) {
       corresponding[key] = [];
     }
     corresponding[key].push(source.sourceId);
-  });
+  }
 
   const returnValue: Record<EntityId, SourceDetails> = {};
 
   const prettyPrintedSources = byKind["prettyPrinted"] || [];
-  prettyPrintedSources.forEach(source => {
+
+  for (let source of prettyPrintedSources) {
     const nonPrettyPrintedVersionId = source.generatedSourceIds![0];
     prettyPrinted.connectNode(nonPrettyPrintedVersionId, source.sourceId);
     canonical.connectNode(source.sourceId, nonPrettyPrintedVersionId);
-  });
-
-  newSources.forEach(source => {
-    returnValue[source.sourceId] = fullSourceDetails({
-      ...omit(source, "sourceId", "generatedSourceIds"),
-      contentHash: contentHashForSource(source),
-      correspondingSourceIds: corresponding[keyForSource(source)],
-      id: source.sourceId,
-      prettyPrinted: prettyPrinted.from(source.sourceId)?.[0],
-      prettyPrintedFrom: prettyPrinted.to(source.sourceId)?.[0],
-      generated: generated.from(source.sourceId) || [],
-      generatedFrom: generated.to(source.sourceId) || [],
-      canonicalId: findCanonicalId(source.sourceId),
-    });
-  });
+  }
+
+  for (let source of newSources) {
+    const { sourceId, generatedSourceIds, ...remainingFields } = source;
+
+    returnValue[sourceId] = fullSourceDetails(
```

Comment:
Why don't we get rid of `fullSourceDetails()`? All it seems to do is fill in some default values which we could do directly here.

Reply:
sure, I'll do that now

Reply:
Hah, ironically all of those _are_ being filled in already :)